File: tests/python/test_data_manage.py
"""Tests for the data array-management operation (rename / drop / keep)."""

import numpy as np
import pytest

import meshioplusplus as mp
from meshioplusplus._data_manage import _manage_py

from .helpers_data import assert_same_geometry, data_mesh


def test_rename_changes_the_key_and_preserves_values():
    m = data_mesh()
    out = mp.data_rename(m, "point", "T", "temperature")
    assert "T" not in out.point_data
    assert "temperature" in out.point_data
    assert np.array_equal(out.point_data["temperature"], m.point_data["T"])
    # Untouched arrays survive.
    assert "v" in out.point_data
    assert "mat" in out.cell_data
    assert "meta" in out.field_data
    assert_same_geometry(m, out)


def test_rename_multi_block_cell_data():
    m = data_mesh()
    out = mp.data_rename(m, "cell", "mat", "material")
    assert "mat" not in out.cell_data
    # One array per cell block must survive.
    assert len(out.cell_data["material"]) == len(m.cells)
    assert out.cell_data["material"][0][0] == 1.0
    assert out.cell_data["material"][1][0] == 3.0


def test_drop_removes_exactly_the_named_arrays():
    m = data_mesh()
    out = mp.data_drop(m, "point", ["T"])
    assert "T" not in out.point_data
    assert "v" in out.point_data
    # Other locations untouched.
    assert set(out.cell_data) == {"mat", "tag"}
    assert "meta" in out.field_data
    assert_same_geometry(m, out)


def test_drop_multiple_names_and_locations():
    m = data_mesh()
    out = mp.data_manage(m, drop=[("point", "T"), ("point", "v"), ("cell", "tag")])[
        "mesh"
    ]
    assert out.point_data == {}
    assert set(out.cell_data) == {"mat"}


def test_keep_retains_only_the_named_subset():
    m = data_mesh()
    out = mp.data_keep(m, "cell", ["tag"])
    assert set(out.cell_data) == {"tag"}
    # A location the whitelist does not mention is untouched.
    assert set(out.point_data) == {"T", "v"}
    assert "meta" in out.field_data


def test_keep_nothing_drops_everything_at_that_location():
    m = data_mesh()
    out = mp.data_keep(m, "point", [])
    assert out.point_data == {}
    assert "mat" in out.cell_data


def test_unknown_key_raises_a_clear_error():
    m = data_mesh()
    with pytest.raises(ValueError) as exc:
        mp.data_drop(m, "point", ["nope"])
    msg = str(exc.value)
    assert "nope" in msg
    assert "point_data" in msg
    # The message lists every available key.
    assert "T" in msg
    assert "v" in msg


def test_ignore_missing_silences_unknown_keys():
    m = data_mesh()
    out = mp.data_drop(m, "point", ["nope"], ignore_missing=True)
    assert "T" in out.point_data


def test_rename_onto_an_existing_name_raises():
    m = data_mesh()
    with pytest.raises(ValueError):
        mp.data_rename(m, "point", "T", "v")


def test_two_renames_to_the_same_target_raise():
    m = data_mesh()
    with pytest.raises(ValueError):
        mp.data_manage(m, rename=[("point", "T", "x"), ("point", "v", "x")])


def test_swapping_two_names_is_allowed():
    m = data_mesh()
    out = mp.data_manage(m, rename=[("point", "T", "v"), ("point", "v", "T")])["mesh"]
    # The old "v" (a 3-vector) is now called "T" and vice versa.
    assert out.point_data["T"].shape == m.point_data["v"].shape
    assert out.point_data["v"].shape == m.point_data["T"].shape


def test_report_lists_what_was_dropped_and_renamed():
    m = data_mesh()
    r = mp.data_manage(m, drop=[("point", "T")], rename=[("field", "meta", "metadata")])
    assert r["dropped"] == ["point_data:T"]
    assert r["renamed"] == [("field_data:meta", "field_data:metadata")]


def test_unknown_location_raises():
    m = data_mesh()
    with pytest.raises(ValueError):
        mp.data_drop(m, "vertex", ["T"])


def test_sets_survive_unchanged():
    m = data_mesh()
    m.point_sets = {"left": np.array([0, 3])}
    m.cell_sets = {"tris": [np.array([0, 1]), np.array([], dtype=int)]}
    out = mp.data_drop(m, "point", ["T"])
    assert np.array_equal(out.point_sets["left"], np.array([0, 3]))
    assert np.array_equal(out.cell_sets["tris"][0], np.array([0, 1]))


def test_cpp_matches_python():
    core = pytest.importorskip("meshioplusplus._core")
    m = data_mesh()
    cases = [
        ([], [("point", "T")], []),
        ([("cell", "mat")], [], []),
        ([], [], [("point", "T", "temp")]),
    ]
    for keep, drop, rename in cases:
        cpp = core.data_manage(m, keep, drop, rename, False)
        py = _manage_py(m, keep, drop, rename, False)
        assert cpp["dropped"] == py["dropped"]
        assert [tuple(t) for t in cpp["renamed"]] == py["renamed"]
        assert sorted(cpp["mesh"].point_data) == sorted(py["mesh"].point_data)
        assert sorted(cpp["mesh"].cell_data) == sorted(py["mesh"].cell_data)


def test_roundtrip_write_read(tmp_path):
    m = data_mesh()
    out = mp.data_rename(m, "point", "T", "temperature")
    path = tmp_path / "out.vtu"
    mp.write(path, out)
    back = mp.read(path)
    assert "temperature" in back.point_data
    assert np.allclose(back.point_data["temperature"], m.point_data["T"])

File: tests/cpp/test_data_manage.cpp
//  ██████   ██████ ██████████  █████████  █████   █████ █████    ███████
// ░░██████ ██████ ░░███░░░░░█ ███░░░░░███░░███   ░░███ ░░███   ███░░░░░███      ███         ███
//  ░███░█████░███  ░███  █ ░ ░███    ░░░  ░███    ░███  ░███  ███     ░░███    ░███        ░███
//  ░███░░███ ░███  ░██████   ░░█████████  ░███████████  ░███ ░███      ░███ ███████████ ███████████
//  ░███ ░░░  ░███  ░███░░█    ░░░░░░░░███ ░███░░░░░███  ░███ ░███      ░███░░░░░███░░░ ░░░░░███░░░
//  ░███      ░███  ░███ ░   █ ███    ░███ ░███    ░███  ░███ ░░███     ███     ░███        ░███
//  █████     █████ ██████████░░█████████  █████   █████ █████ ░░░███████░      ░░░         ░░░
// ░░░░░     ░░░░░ ░░░░░░░░░░  ░░░░░░░░░  ░░░░░   ░░░░░ ░░░░░    ░░░░░░░
//
//
//  License:         MIT License
//                   meshio++ default license: LICENSE
//
//  Main authors:    Vicente Mataix Ferrandiz
//
//
// Tests for the data array-management operation (rename / drop / keep).

// System includes
#include <stdexcept>
#include <string>
#include <vector>

// External includes
#include <gtest/gtest.h>

// Project includes
#include "mesh_fixtures.hpp"
#include "meshioplusplus/detail/value_io.hpp"
#include "meshioplusplus/operations/data_manage.hpp"

namespace {

using meshioplusplus::DataKey;
using meshioplusplus::DataLocation;
using meshioplusplus::DataManageOptions;
using meshioplusplus::DataManageResult;
using meshioplusplus::DataRename;
using meshioplusplus::Mesh;

TEST(DataManage, RenamePreservesValuesAndGeometry) {
    Mesh in = mt::data_mesh();
    Mesh out = meshioplusplus::data_rename(in, DataLocation::Point, "T", "temperature");

    EXPECT_FALSE(out.HasPointData("T"));
    ASSERT_TRUE(out.HasPointData("temperature"));
    ASSERT_EQ(out.PointData("temperature").Size(), in.PointData("T").Size());
    for (std::size_t i = 0; i < in.PointData("T").Size(); ++i)
        EXPECT_DOUBLE_EQ(meshioplusplus::detail::read_double(out.PointData("temperature"), i),
                         meshioplusplus::detail::read_double(in.PointData("T"), i));
    // Untouched arrays survive.
    EXPECT_TRUE(out.HasPointData("v"));
    EXPECT_TRUE(out.HasCellData("mat"));
    EXPECT_TRUE(out.HasFieldData("meta"));
    mt::expect_same_geometry(in, out);
}

TEST(DataManage, RenameMultiBlockCellData) {
    Mesh in = mt::data_mesh();
    Mesh out = meshioplusplus::data_rename(in, DataLocation::Cell, "mat", "material");
    ASSERT_TRUE(out.HasCellData("material"));
    EXPECT_FALSE(out.HasCellData("mat"));
    // Every block must survive the rename — the uniform API requires exactly
    // one cell_data array per cell block.
    ASSERT_EQ(out.CellDataNumBlocks("material"), out.NumCellBlocks());
    EXPECT_DOUBLE_EQ(meshioplusplus::detail::read_double(out.CellData("material", 0), 0), 1.0);
    EXPECT_DOUBLE_EQ(meshioplusplus::detail::read_double(out.CellData("material", 1), 0), 3.0);
}

TEST(DataManage, DropRemovesExactlyTheNamedArrays) {
    Mesh in = mt::data_mesh();
    Mesh out = meshioplusplus::data_drop(in, DataLocation::Point, {"T"});
    EXPECT_FALSE(out.HasPointData("T"));
    EXPECT_TRUE(out.HasPointData("v"));
    // Other locations untouched.
    EXPECT_TRUE(out.HasCellData("mat"));
    EXPECT_TRUE(out.HasCellData("tag"));
    EXPECT_TRUE(out.HasFieldData("meta"));
    mt::expect_same_geometry(in, out);
}

TEST(DataManage, KeepRetainsOnlyTheNamedSubset) {
    Mesh in = mt::data_mesh();
    Mesh out = meshioplusplus::data_keep(in, DataLocation::Cell, {"tag"});
    EXPECT_TRUE(out.HasCellData("tag"));
    EXPECT_FALSE(out.HasCellData("mat"));
    // A location the whitelist does not mention is left completely alone.
    EXPECT_TRUE(out.HasPointData("T"));
    EXPECT_TRUE(out.HasPointData("v"));
    EXPECT_TRUE(out.HasFieldData("meta"));
}

TEST(DataManage, KeepNothingDropsEverythingAtThatLocation) {
    Mesh in = mt::data_mesh();
    Mesh out = meshioplusplus::data_keep(in, DataLocation::Point, {});
    EXPECT_EQ(out.NumPointData(), 0u);
    EXPECT_TRUE(out.HasCellData("mat"));
}

TEST(DataManage, UnknownKeyListsAvailableKeys) {
    Mesh in = mt::data_mesh();
    try {
        meshioplusplus::data_drop(in, DataLocation::Point, {"nope"});
        FAIL() << "expected an exception";
    } catch (const std::invalid_argument& e) {
        const std::string msg = e.what();
        EXPECT_NE(msg.find("nope"), std::string::npos);
        EXPECT_NE(msg.find("point_data"), std::string::npos);
        // Every available key is named.
        EXPECT_NE(msg.find("T"), std::string::npos);
        EXPECT_NE(msg.find("v"), std::string::npos);
    }
}

TEST(DataManage, IgnoreMissingSilencesUnknownKeys) {
    Mesh in = mt::data_mesh();
    Mesh out = meshioplusplus::data_drop(in, DataLocation::Point, {"nope"}, /*IgnoreMissing=*/true);
    EXPECT_TRUE(out.HasPointData("T"));
}

TEST(DataManage, RenameOntoExistingNameThrows) {
    Mesh in = mt::data_mesh();
    EXPECT_THROW(meshioplusplus::data_rename(in, DataLocation::Point, "T", "v"),
                 std::invalid_argument);
}

TEST(DataManage, TwoRenamesToTheSameTargetThrow) {
    Mesh in = mt::data_mesh();
    DataManageOptions opts;
    opts.rename.push_back(DataRename{DataLocation::Point, "T", "x"});
    opts.rename.push_back(DataRename{DataLocation::Point, "v", "x"});
    EXPECT_THROW(meshioplusplus::data_manage(in, opts), std::invalid_argument);
}

TEST(DataManage, SwapNamesIsAllowed) {
    // T -> v and v -> T: neither target "already exists" in the result, because
    // both are renamed away in the same pass.
    Mesh in = mt::data_mesh();
    DataManageOptions opts;
    opts.rename.push_back(DataRename{DataLocation::Point, "T", "v"});
    opts.rename.push_back(DataRename{DataLocation::Point, "v", "T"});
    DataManageResult r = meshioplusplus::data_manage(in, opts);
    ASSERT_TRUE(r.mMesh.HasPointData("T"));
    ASSERT_TRUE(r.mMesh.HasPointData("v"));
    // The old "v" (a 3-vector) is now called "T".
    EXPECT_EQ(r.mMesh.PointData("T").Size(), in.PointData("v").Size());
    EXPECT_EQ(r.mMesh.PointData("v").Size(), in.PointData("T").Size());
}

TEST(DataManage, ReportsWhatWasDroppedAndRenamed) {
    Mesh in = mt::data_mesh();
    DataManageOptions opts;
    opts.drop.push_back(DataKey{DataLocation::Point, "T"});
    opts.rename.push_back(DataRename{DataLocation::Field, "meta", "metadata"});
    DataManageResult r = meshioplusplus::data_manage(in, opts);
    ASSERT_EQ(r.mDropped.size(), 1u);
    EXPECT_EQ(r.mDropped[0], "point_data:T");
    ASSERT_EQ(r.mRenamed.size(), 1u);
    EXPECT_EQ(r.mRenamed[0].first, "field_data:meta");
    EXPECT_EQ(r.mRenamed[0].second, "field_data:metadata");
}

TEST(DataManage, KeepThenDropThenRenameOrder) {
    Mesh in = mt::data_mesh();
    DataManageOptions opts;
    opts.keep.push_back(DataKey{DataLocation::Cell, "mat"});
    opts.keep.push_back(DataKey{DataLocation::Cell, "tag"});
    opts.drop.push_back(DataKey{DataLocation::Cell, "tag"});
    opts.rename.push_back(DataRename{DataLocation::Cell, "mat", "material"});
    DataManageResult r = meshioplusplus::data_manage(in, opts);
    EXPECT_TRUE(r.mMesh.HasCellData("material"));
    EXPECT_FALSE(r.mMesh.HasCellData("tag"));
    EXPECT_FALSE(r.mMesh.HasCellData("mat"));
}

}  // namespace
